### PR TITLE
texture manager fixes

### PR DESCRIPTION
**Pull request by rogueSkib**
* changed texture manager algorithm to scale memory limits
* fixed gl error handling that caused spamming and crashing on android
* fixed half sized textures
  * half sized textures are avoided unless necessary

### CODE REVIEW

File: core.c [hunk 1]
```diff
@@ -336,29 +330,28 @@ void core_reset() {
 /**
  * @name	core_check_gl_error
  * @brief	check's for a gl error and reports it to the platform
- * @retval	NONE
+ * @retval	(bool) whether or not there was an error
  */
-void core_check_gl_error() {
+bool core_check_gl_error() {
```
Comment (jwilm) [inline core.c]: The function name here is really misleading. It says, in english, check if there was a GL error and tell me yes or no. However, there are a bunch of side effects in here. Maybe an in-out param can be added which is set to the error code if there was an error or -1 otherwise. The caller is then responsible for checking if it was a memory warning or whatever else.

```c
int gl_error;
if (core_check_gl_error(&gl_error)) {
  switch gl_error:
    case GLERROR_OUT_OF_MEMORY:
      // ...
      break;
    // ...
}
```

File: texture_manager.c [hunk 2]
```diff
@@ -985,13 +870,12 @@ void texture_manager_tick(texture_manager *manager) {
  * textures.  The texture load loop may happen, but it will resolve itself
  * quickly by reloading textures half-sized.
  */
-void texture_manager_memory_warning(texture_manager *manager) {
```
Comment (jwilm) [inline texture_manager.c]: It seems like this should continue to take a `texture_manager*` to match the rest of the methods. I think the idea here is not having global state. `m_memory_warning` should be a property on the `texture_manager`.

File: texture_manager.h [hunk 3]
```diff
@@ -55,16 +55,16 @@ void texture_manager_destroy(texture_manager *manager);
 void texture_manager_clear_textures(texture_manager *manager, bool clear_all);
 void texture_manager_free_texture(texture_manager *manager, texture_2d *tex);
 void texture_manager_touch_texture(texture_manager *manager, const char *url);
-void texture_manager_set_use_halfsized_textures();
```
Comment (jwilm) [inline texture_manager.h]: Similarly here, we should take a `texture_manager*` and have a `use_halfsized` flag on the texture_manager itself.

File: texture_manager.c [hunk 4]
```diff
@@ -31,44 +31,23 @@
...
 #define DEFAULT_SHEET_DIMENSION 64
 
-// Constants for Android phones and starting-points for iOS
-#define MAX_BYTES_FOR_HALFSIZED_TEXTURES 500000000  /* 500 MB formerly 30 MB */
-#define MAX_BYTES_FOR_FULLSIZED_TEXTURES 500000000  /* 500 MB formerly 100 MB */
-
-// Does not use fullsized textures under this memory value
-#define MIN_BYTES_FOR_FULLSIZE_TEXTURES 30000000    /* 30 MB */
+// Large number to start texture memory limit
+#define MAX_BYTES_FOR_TEXTURES 500000000    /* 500 MB */
```
Comment (jwilm) [inline texture_manager.c]: Since this supports ratcheting upwards, maybe we should start with 100MB and let it grow as needed (assuming all else is equivalent).

File: texture_manager.c [hunk 5]
```diff
@@ -410,78 +355,64 @@ static int last_accessed_compare(texture_2d *a, texture_2d *b) {
...
+#endif
+
 }
 
 void texture_manager_reload_canvases(texture_manager *manager) {
     texture_2d *tex = NULL;
     texture_2d *tmp = NULL;
     HASH_ITER(url_hash, manager->url_to_tex, tex, tmp) {
-        if (tex->is_canvas || tex->is_text) {
+        if (tex->is_canvas) {
```
Comment (jwilm) [inline texture_manager.c]: Why doesn't this code care about `tex->is_text` now?

File: texture_manager.c [hunk 6]
```diff
@@ -594,17 +519,31 @@ void texture_manager_background_texture_loader(void *dummy) {
...
         while (cur_tex) {
             texture_2d *old_cur = NULL;
-
-            if (cur_tex->pixel_data == NULL && cur_tex->url != NULL && !cur_tex->failed) {
-                TEXLOG("Passing to load_image_with_c: %s", cur_tex->url);
-
-                if (!resource_loader_load_image_with_c(cur_tex)) { //if not loading from C remove from list
+            const char *url = cur_tex->url;
+
+            if (url != NULL) {
```
Comment (jwilm) [inline texture_manager.c]: This code to check whether this is a canvas is in two places. There should be a `static inline bool` function to check whether a `const char*` says `__canvas__`.

File: texture_manager.c (hunk 4 above)
Comment (rogueSkib) [inline texture_manager.c]: We want to have as many images in memory at possible to avoid lag spikes from having to load them during gameplay, so having this start high seems ideal. It throws out textures as soon as this limit is reached, and even though it grows, it still has to let go of textures.

File: texture_manager.c (hunk 5 above)
Comment (rogueSkib) [inline texture_manager.c]: The is_text flag was not being set to true ever. This logic was left over from a time when text needed to be treated like canvases.

Comment (rogueSkib) [inline texture_manager.c]: I correctly set the is_text flag to true, and it broke in places like this.

File: texture_manager.c [hunk 7]
```diff
@@ -229,20 +189,6 @@ void texture_manager_get_sheet_size(char *url, int *width, int *height) {
         }
     }
 
-    if (fontsheet_map_root) {
```
Comment (jwilm) [inline texture_manager.c]: Why was this removed?

File: core.c (hunk 1 above)
Comment (rogueSkib) [inline core.c]: This is a good suggestion - I don't have time to refactor yet though - because there are a lot of similar changes I would make. These fixes focus on working logic so we can ship safely on all devices. Let's circle back and optimize / refactor when we have time!

File: texture_manager.c (hunk 6 above)
Comment (rogueSkib) [inline texture_manager.c]: Agreed, that would be a good refactor.

File: texture_manager.c (hunk 7 above)
Comment (rogueSkib) [inline texture_manager.c]: This was deprecated font sheet logic that's not longer supported in devkit and caused lots of errors to be emitted for no reason.

File: texture_manager.c (hunk 5 above)
Comment (jwilm) [inline texture_manager.c]: So I guess the real question is then, why doesn't text need to be treated like a canvas now, or did it ever?

File: core.c (hunk 1 above)
Comment (jwilm) [inline core.c]: Alright. Would you mind making a few issues for such items?

**Comment by jwilm** [comment]
Seems good overall. As long as this has been tested on several devices with a couple of games, I think it's fine to merge.

File: texture_manager.c (hunk 5 above)
Comment (rogueSkib) [inline texture_manager.c]: It did at one points - our text rendering has gone through a lot of changes over the years :)

It's implemented differently per platform, see here for the android implementation where the flag is now being set:

https://github.com/gameclosure/native-android/blob/master/TeaLeaf/jni/platform/text_manager.cpp

Comment (rogueSkib) [inline texture_manager.c]: Here's the pull request change that's relevant:

https://github.com/gameclosure/native-android/pull/18/files#diff-13049a9e9c485f0825731837ca1ed5b3R60
